README.md
thief: Temporal HIErarchical Forecasting <img src="man/figures/logo.png" align="right" />
======================

[![Travis-CI Build Status](https://travis-ci.org/robjhyndman/thief.svg?branch=master)](https://travis-ci.org/robjhyndman/thief)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/thief)](https://cran.r-project.org/package=thief)
[![Downloads](http://cranlogs.r-pkg.org/badges/thief)](https://cran.r-project.org/package=thief)
[![Licence](https://img.shields.io/badge/licence-GPL--3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0.en.html)

The R package *thief* provides methods and tools for generating forecasts at different temporal frequencies using a hierarchical time series approach.

Authors: Rob J Hyndman and Nikolaos Kourentzes

This package implements the methods described in

[Athanasopoulos, G., Hyndman, R.J., Kourentzes, N., and Petropoulos, F. (2016) Forecasting with temporal hierarchies.](http://robjhyndman.com/publications/temporal-hierarchies/)


## Installation
You can install the **stable** version on
[R CRAN](https://cran.r-project.org/package=thief).

```s
install.packages('thief', dependencies = TRUE)
```

You can install the **development** version from
[Github](https://github.com/robjhyndman/thief)

```s
# install.packages("devtools")
devtools::install_github("robjhyndman/thief")
```

## Usage

```s
library(thief)
thief(USAccDeaths)
```

## License

This package is free and open source software, licensed under GPL 3.


NEWS.md
## V 0.3 (24 January 2018)
  * Improved handling of forecastfunction argument in thief()
  * Bug fixes
  * pkgdown documentation

## V 0.2 (7 September 2016)
  * First version on CRAN

## V 0.1 (22 August 2016)
  * First version on github


R/NHSdata.R
#' Accident and Emergency demand in the UK
#'
#' Weekly demand of Accident & Emergency departments in the UK, 
#' from 7 November 2010 to 7 June 2015.
#'
#' @docType data
#'
#' @usage AEdemand
#'
#' @format An object of class \code{ts}.
#'
#' @keywords datasets
#'
#' @examples
#' library(ggplot2)
#' autoplot(AEdemand, xlab="Year", ylab="Demand ('000)") +
#'   ggtitle("Accident & Emergency Demand in the UK")
#' 
#' \dontrun{
#' 
#' # Demonstration of the adjustment of all temporal aggregates
#' # using Total Emergency Admissions
#' 
#' total <- AEdemand[,12]
#' totalagg <- tsaggregates(total)
#' plot(totalagg, main="Total Emergency Admissions")
#' 
#' # Base forecasts
#' base <- list()
#' for(i in 1:5)
#'   base[[i]] <- forecast(auto.arima(totalagg[[i]]))
#' base[[6]] <-  forecast(auto.arima(totalagg[[6]]), h=2)
#'
#' # Reconciled forecasts
#' reconciled <- reconcilethief(base)
#'
#' main <- paste(names(totalagg)," (k=",
#'            52/unlist(lapply(reconciled,frequency)),")",sep="")
#' par(mfrow=c(2,3))
#' for(i in 6:1)
#' {
#'   ylim <- range(base[[i]]$mean, base[[i]]$x, reconciled[[i]]$mean)
#'   plot(base[[i]], main=main[i], fcol='white',
#'       plot.conf=FALSE, ylim=ylim, xlim=c(2010.5,2017.5))
#'   polygon(c(2015.45, 2020, 2020, 2015.45),
#'           c(0, 0, 1e5, 1e5), col='grey', border=FALSE)
#'   lines(base[[i]]$mean, col='red', lty=2)
#'   lines(reconciled[[i]]$mean, col='blue')
#' }
#' }
"AEdemand"



R/combine.R
#' Reconcile temporal hierarchical forecasts
#'
#' Takes forecasts of time series at all levels of temporal aggregation
#' and combines them using the temporal hierarchical approach of Athanasopoulos et al (2016).
#'
#' @param forecasts List of forecasts. Each element must be a time series of forecasts,
#' or a forecast object.
#' The number of forecasts should be equal to k times the seasonal period for each series,
#' where k is the same across all series.
#' @param comb Combination method of temporal hierarchies, taking one of the following values:
#' \describe{
#'   \item{"struc"}{Structural scaling - weights from temporal hierarchy}
#'   \item{"mse"}{Variance scaling - weights from in-sample MSE}
#'   \item{"ols"}{Unscaled OLS combination weights}
#'   \item{"bu"}{Bottom-up combination -- i.e., all aggregate forecasts are ignored.}
#'   \item{"shr"}{GLS using a shrinkage (to block diagonal) estimate of residuals}
#'   \item{"sam"}{GLS using sample covariance matrix of residuals}
#' }
#' @param mse A vector of one-step MSE values corresponding to each of the forecast series.
#' @param residuals List of residuals corresponding to each of the forecast models.
#' Each element must be a time series of residuals. If \code{forecast} contains a list of
#' forecast objects, then the residuals will be extracted automatically and this argument
#' is not needed. However, it will be used if not \code{NULL}.
#' @param returnall If \code{TRUE}, a list of time series corresponding to the first argument
#' is returned, but now reconciled. Otherwise, only the most disaggregated series is returned.
#' @param aggregatelist (optional) User-selected list of forecast aggregates to consider
#'
#' @return
#'   List of reconciled forecasts in the same format as \code{forecast}.
#' If \code{returnall==FALSE}, only the most disaggregated series is returned.
#' @seealso \code{\link{thief}}, \code{\link{tsaggregates}}
#'
#' @examples
#' # Construct aggregates
#' aggts <- tsaggregates(USAccDeaths)
#'
#' # Compute forecasts
#' fc <- list()
#' for(i in seq_along(aggts))
#'   fc[[i]] <- forecast(auto.arima(aggts[[i]]), h=2*frequency(aggts[[i]]))
#'
#' # Reconcile forecasts
#' reconciled <- reconcilethief(fc)
#'
#' # Plot forecasts before and after reconcilation
#' par(mfrow=c(2,3))
#' for(i in seq_along(fc))
#' {
#'   plot(reconciled[[i]], main=names(aggts)[i])
#'   lines(fc[[i]]$mean, col='red')
#' }
#'
#' @export
#' @author Rob J Hyndman

reconcilethief <- function(forecasts,
               comb=c("struc","mse","ols","bu","shr","sam"),
               mse=NULL, residuals=NULL, returnall=TRUE,
               aggregatelist=NULL)
{
  comb <- match.arg(comb)

  # If forecasts is a list of forecast objects, then
  # extract list of forecast time series and list of residual time series
  if(is.element("forecast",class(forecasts[[1]])))
  {
    returnclass <- "forecast"
    origf <- forecasts
    # Grab residuals
    if(is.null(residuals))
    {
      residuals <- list()
      for(i in seq_along(forecasts))
        residuals[[i]] <- residuals(forecasts[[i]])
      # Discard partial years at start of residual series
      for(i in seq_along(residuals))
      {
        tspy <- tsp(residuals[[i]])
        m <- tspy[3]
        fullyears <- trunc(length(residuals[[i]])/m)
        residuals[[i]] <- ts(utils::tail(residuals[[i]], fullyears*m),
          frequency=m, end=tspy[2])
      }
    }
    # Grab forecasts
    for(i in seq_along(forecasts))
      forecasts[[i]] <- forecasts[[i]]$mean
  }
  else
    returnclass <- "ts"

  # Find seasonal periods
  freq <- unlist(lapply(forecasts,frequency))
  if(min(freq) != 1L)
    stop("Minimum seasonal period should be 1")
  m <- max(freq)
  # Put in order
  k <- rev(order(freq))
  forecasts <- forecasts[k]
  freq <- freq[k]

  # Check series each have same equivalent lengths.
  lengths <- unlist(lapply(forecasts, length))
  if(!is.constant(lengths/freq))
    stop("Forecast series must have the same equivalent lengths")

  # Bottom up
  if(comb == "bu")
    bts <- forecasts[[1]]
  # Some combination
  else
  {
    # Set up group matrix for hts
    # (adjusted to allow consideration of aggregatelist input)
    nsum <- rev(rep(m/freq, freq))
    unsum <- unique(nsum)
    grps <- matrix(0, nrow=length(unsum)-1, ncol=m)
    for(i in 1:(length(unsum)-1))
    {
      mi <- m/unsum[i]
      grps[i,] <- rep(1:mi, rep(unsum[i],mi))
    }

    # Set up matrix of forecasts in right structure
    nc <- length(forecasts[[1]])/m
    fmat <- matrix(0, nrow=0, ncol=nc)
    for(i in rev(seq_along(forecasts)))
      fmat <- rbind(fmat, matrix(forecasts[[i]], ncol=nc))

    # OLS or WLS reconciliation
    if(is.element(comb, c("struc","ols","mse")))
    {
      if(comb=="struc")
        weights <- 1/nsum
      else if(comb=="ols")
        weights <- NULL
      else if(comb=="mse")
        weights <- 1/rep(rev(mse), rev(unsum))
      bts <- hts::combinef(t(fmat), groups=grps, weights=weights, keep='bottom')
    }
    else   # GLS reconciliation
    {
      # Set up matrix of residuals in right structure
      if(is.null(residuals))
        stop("GLS needs residuals")
      nc <- length(residuals[[1]])/m
      rmat <- matrix(0, nrow=0, ncol=nc)
      for(i in rev(seq_along(forecasts)))
        rmat <- rbind(rmat, matrix(residuals[[i]], ncol=nc))
      bts <- hts::MinT(t(fmat), groups=grps, residual=t(rmat),
        covariance=comb, keep='bottom')
    }
    # Turn resulting reconciled forecasts back into a ts object
    bts <- ts(c(t(bts)))
    tsp(bts) <- tsp(forecasts[[1]])
  }

  # Now figure out what to return
  if(returnclass=="ts") # Just return time series
  {
    if(!returnall)
      return(bts)
    else
      return(tsaggregates(bts, aggregatelist=aggregatelist))
  }
  else #return forecast objects
  {
    if(!returnall)
    {
      adj <- bts - origf[[1]]$mean
      origf[[1]]$mean <- bts
      if(!is.null(origf[[1]]$lower))
      {
        origf[[1]]$lower <- sweep(origf[[1]]$lower,1,adj,"+")
        origf[[1]]$upper <- sweep(origf[[1]]$upper,1,adj,"+")
      }
      return(origf[[1]])
    }
    else
    {
      allts <- tsaggregates(bts, aggregatelist=aggregatelist)
      for(i in seq_along(origf))
      {
        adj <- allts[[i]] - origf[[i]]$mean
        origf[[i]]$mean <- allts[[i]]
        if(!is.null(origf[[i]]$lower))
        {
          origf[[i]]$lower <- sweep(origf[[i]]$lower,1,adj,"+")
          origf[[i]]$upper <- sweep(origf[[i]]$upper,1,adj,"+")
        }
      }
      return(origf)
    }
  }
}


R/tsaggregates.R
#' Non-overlapping temporal aggregation of a time series
#'
#' Produces all temporal aggregations for frequencies greater than 1
#'
#' @param y Univariate time series of class \code{ts}.
#' @param m Integer seasonal period
#' @param align Indicates how the aggregates are to be aligned:
#' either with the \code{start} of the series or the \code{end} of the series.
#' For forecasting purposes, it should be set to \code{end}.
#' @param aggregatelist User-selected list of aggregates to consider.
#'
#' @return A list of time series. The first element is the series `y`,
#' followed by series with increasing levels of aggregation. The last
#' element is the "annual" series (i.e., the series aggregated over all seasons).
#' @seealso \code{\link{plot.tsaggregates}}
#'
#' @examples
#' tsaggregates(USAccDeaths)
#'
#' @author Rob J Hyndman
#' @export


tsaggregates <- function (y, m=frequency(y), align=c("end","start"),
                          aggregatelist=NULL)
{
  align <- match.arg(align)
  n <- length(y)
  m <- as.integer(m)

  # Find all factors of m
  mout <- seq_len(m)
  mout <- mout[m %% mout == 0L]
  mout <- mout[mout <= n]
  if (length(mout) == 0L)
    stop("Series too short for aggregation")

  # If user has specified aggregatelist, then consider only those aggregates
  if (!is.null(aggregatelist))
    mout <- mout[mout %in% aggregatelist]
  if (length(mout) == 0L)
    stop("No valid factors in aggregatelist argument")

  k <- length(mout)
  y.out <- vector("list",k)
  y.out[[1L]] <- y
  if(!is.ts(y))
    y <- ts(y, frequency=m)
  for(i in seq_len(k)[-1L])
  {
    if(align=='end')
      start <- n%%mout[i] + 1L
    else
      start <- 1L
    nk <- trunc(n/mout[i])
    tmp <- matrix(y[start - 1L + seq_len(mout[i]*nk)], ncol=nk)
    y.out[[i]] <- ts(colSums(tmp), frequency=m/mout[i],
      start=tsp(y)[1] + (start-1)/m)
  }
  names(y.out) <- paste("Period", m/mout)
  # Give names to common periods
  if(m==4L)
  {
    names(y.out)[mout==4L] <- "Annual"
    names(y.out)[mout==2L] <- "Biannual"
    names(y.out)[mout==1L] <- "Quarterly"
  }
  else if(m == 12L)
  {
    names(y.out) <- paste(mout,"-Monthly",sep="")
    names(y.out)[mout==12L] <- "Annual"
    names(y.out)[mout==6L] <- "Biannual"
    names(y.out)[mout==3L] <- "Quarterly"
    names(y.out)[mout==1L] <- "Monthly"
  }
  else if(m == 7L)
  {
    names(y.out)[mout==7L] <- "Weekly"
    names(y.out)[mout==1L] <- "Daily"
  }
  else if(m == 24L | m == 168L | m == 8760L)
  {
    names(y.out) <- paste(mout,"-Hourly",sep="")
    j <- mout%%24L == 0L
    names(y.out)[j] <- paste(mout[j]/24L,"-Daily",sep="")
    j <- mout%%168L == 0L
    names(y.out)[j] <- paste(mout[j]/168L,"-Weekly",sep="")
    j <- mout%%8760L == 0L
    names(y.out)[j] <- paste(mout[j]/8760L,"-Yearly",sep="")
    names(y.out)[mout==8760L] <- "Annual"
    names(y.out)[mout==2190L] <- "Quarterly"
    names(y.out)[mout==168L] <- "Weekly"
    names(y.out)[mout==24L] <- "Daily"
    names(y.out)[mout==1L] <- "Hourly"
  }
  else if(m == 48L | m == 336L | m == 17520L)
  {
    j <- mout%%2L == 0L
    names(y.out)[j] <- paste(mout[j]/2L,"-Hourly",sep="")
    j <- mout%%48L == 0L
    names(y.out)[j] <- paste(mout[j]/48L,"-Daily",sep="")
    j <- mout%%336L == 0L
    names(y.out)[j] <- paste(mout[j]/336L,"-Weekly",sep="")
    j <- mout%%17520L == 0L
    names(y.out)[j] <- paste(mout[j]/17520L,"-Yearly",sep="")
    names(y.out)[mout==17520L] <- "Annual"
    names(y.out)[mout==4380L] <- "Quarterly"
    names(y.out)[mout==336L] <- "Weekly"
    names(y.out)[mout==48L] <- "Daily"
    names(y.out)[mout==2L] <- "Hourly"
    names(y.out)[mout==1L] <- "Half-hourly"
  }
  else if(m == 52L)
  {
    names(y.out) <- paste(mout,"-Weekly",sep="")
    names(y.out)[mout==52L] <- "Annual"
    names(y.out)[mout==26L] <- "Biannual"
    names(y.out)[mout==13L] <- "Quarterly"
    names(y.out)[mout==1L] <- "Weekly"
  }
  return(structure(y.out, class="tsaggregates"))
}


R/plot.tsaggregates.R
#' Plot time series aggregates
#'
#' Plots all temporal aggregations of a time series
#'
#' @param x \code{tsaggregates} object, produced by \code{\link{tsaggregates}}.
#' @param object \code{tsaggregates} object, produced by \code{\link{tsaggregates}}.
#' @param series The indexes of the series to plot. By default, all series are plotted.
#' @param ... Other arguments passed to \code{\link[stats]{plot.ts}}
#' or \code{\link[forecast]{autoplot.ts}}.
#'
#' @method plot tsaggregates
#'
#' @examples
#' deathagg <- tsaggregates(USAccDeaths)
#' plot(deathagg, series=c(1,2,4,6))
#'
#' library(ggplot2)
#' autoplot(deathagg)
#'
#' @author Rob J Hyndman
#' @export
#' @importFrom stats plot.ts
#' @importFrom graphics plot

plot.tsaggregates <- function(x, series="all", ...)
{
  tmp <- mts.from.tsaggregates(x,series,...)
  args <- tmp$args
  args$x <- tmp$tsmat
  if(!is.element("main",names(args)))
    args$main <- deparse(substitute(x))
  if(length(tmp$series)==1L & !is.element("ylab",names(args)))
    args$ylab <- names(x)[tmp$series]
  if(length(tmp$series) <= 6L & !is.element("nc",names(args)))
    args$nc <- 1L
  do.call(plot.ts, args)
}

#' @importFrom ggplot2 autoplot
#' @rdname plot.tsaggregates
#' @method autoplot tsaggregates
#' @export

autoplot.tsaggregates <- function(object, series="all", ...)
{
  tmp <- mts.from.tsaggregates(object,series,...)
  args <- tmp$args
  args$object <- tmp$tsmat
  args$facets <- TRUE
  if(!is.element("main",names(args)))
    args$main <- deparse(substitute(object))
  if(!is.element("ylab",names(args)))
  {
    if(length(tmp$series)==1L)
      args$ylab <- names(object)[tmp$series]
    else
      args$ylab <- ""
  }
  do.call(ggplot2::autoplot, args)
}

#' @importFrom stats approx frequency is.ts time ts tsp "tsp<-"

mts.from.tsaggregates <- function(x, series, ...)
{
  nseries <- length(x)

  if(identical(series,"all"))
    series <- seq_len(nseries)
  if(length(series) > 10L)
    stop("Too many series to plot. Please select a subset.")
  if(min(series) < 1L | max(series) > nseries)
    stop(paste("series should be between 1 and",nseries))

  args <- list(...)

  # Create linearly interpolated mts object
  times <- time(x[[1]])
  tsmat <- matrix(0, nrow=length(times), ncol=nseries)

  tsmat[,1] <- x[[1]]
  if(nseries > 1L)
  {
    for(j in 2:nseries)
    {
      if(length(x[[j]]) > 1L)
        tsmat[,j] <- approx(time(x[[j]]), x[[j]], xout=times)$y
    }
  }
  tsmat <- ts(tsmat)
  tsp(tsmat) <- tsp(x[[1]])
  colnames(tsmat) <- names(x)
  tsmat <- tsmat[,series]

  if(!is.element("xlab",names(args)))
  {
    m <- frequency(x[[1]])
    if(m == 12L | m==4L | m==8760L | m==17520L)
      args$xlab <- "Year"
    else if(m==7L)
      args$xlab <- "Week"
    else if(m==24L | m==48L)
      args$xlab <- "Day"
    else if(m==168L | m==336L)
      args$xlab <- "Week"
    else
      args$xlab <- "Time"
  }

  return(list(tsmat=tsmat, args=args, series=series))
}


R/thief.R
#' Temporal hierarchical forecasting
#'
#' Takes a time series as input and produces forecasts using
#' the temporal hierarchical approach of Athanasopoulos et al (2016).
#'
#' @param y        Time series input
#' @param m        Seasonal period
#' @param h        Forecast horizon
#' @param comb     Combination method of temporal hierarchies, taking one of the following values:
#' \describe{
#'   \item{"struc"}{Structural scaling - weights from temporal hierarchy}
#'   \item{"mse"}{Variance scaling - weights from in-sample MSE}
#'   \item{"ols"}{Unscaled OLS combination weights}
#'   \item{"bu"}{Bottom-up combination -- i.e., all aggregate forecasts are ignored.}
#'   \item{"shr"}{GLS using a shrinkage (to block diagonal) estimate of residuals}
#'   \item{"sam"}{GLS using sample covariance matrix of residuals}
#' }
#' @param usemodel    Model used for forecasting each aggregation level:
#' \describe{
#'   \item{"ets"}{exponential smoothing, using the \code{\link[forecast]{ets}} function.}
#'   \item{"arima"}{arima, using the \code{\link[forecast]{auto.arima}} function.}
#'   \item{"theta"}{theta method, using the \code{\link[forecast]{thetaf}} function.}
#'   \item{"naive"}{random walk forecasts}
#'   \item{"snaive"}{seasonal naive forecasts, based on the last year of observed data.}
#' }
#' @param forecastfunction User-defined function to be used instead of \code{usemodel}. The
#' function must take a time series as the first argument, and the forecast horizon
#' as the second argument. It must return an object of class \code{forecast}.
#' @param aggregatelist User-selected list of forecast aggregates to consider
#' @param ...   Arguments to be passed to the time series modelling function
#' (such as \code{ets} or \code{auto.arima}), or to \code{forecastfunction}.
#'
#' @details This function computes the temporal aggregates of \code{y} using
#' \code{\link{tsaggregates}}, then calculates all forecasts using the model function
#' specified by \code{usemodel} or \code{forecastfunction}, and finally reconciles the
#' forecasts using \code{\link{reconcilethief}}. The reconciled forecasts of \code{y}
#' are returned.
#'
#' @return
#'   forecast object.
#'
#' @seealso \code{\link{reconcilethief}}
#'
#' @examples
#' \dontrun{
#'
#' # Select ARIMA models for all series using auto.arima()
#' z <- thief(AEdemand[,12], usemodel='arima')
#' plot(z)
#'
#' # Use your own function
#' ftbats <- function(y,h,...){forecast(tbats(y),h,...)}
#' z <- thief(AEdemand[,12], forecastfunction=ftbats)
#' plot(z)
#' }
#'
#' @export
#' @import forecast
#' @author Rob J Hyndman and Nikolaos Kourentzes

thief <- function(y, m=frequency(y), h=m*2,
               comb=c("struc","mse","ols","bu","shr","sam"),
               usemodel=c("ets","arima","theta","naive","snaive"),
               forecastfunction=NULL,
               aggregatelist=NULL, ...)
{
  comb <- match.arg(comb)
  if(is.null(forecastfunction))
    usemodel <- match.arg(usemodel)
  else
    usemodel <- deparse(substitute(forecastfunction))

  # Check input is a univariate time series
  if(!is.element("ts",class(y)))
    stop("y must be a time series object")
  if(NCOL(y) > 1L)
    stop("y must be a univariate time series")

  # Make sure the time series is seasonal
  if(m <= 1L)
    stop("Seasonal period (m) must be greater than 1")
  if(length(y) < 2*m)
    stop("I need at least 2 periods of data")

  # Check validity of aggregatelist argument
  if(!is.null(aggregatelist))
    if(!((1 %in% aggregatelist)&&(m %in% aggregatelist)))
      stop("aggregatelist must include most disaggregate level (seasonal period m)
           AND most aggregated level (with seasonal period 1)")

  # Compute aggregate
  aggy <- tsaggregates(y, aggregatelist=aggregatelist)

  # Compute forecasts
  frc <- th.forecast(aggy, h=h, usemodel=usemodel,
    forecastfunction=forecastfunction, ...)

  # Reconcile forecasts and fitted values
  bts <- reconcilethief(frc$forecast, comb, frc$mse, frc$residuals,
    returnall=FALSE, aggregatelist=aggregatelist)
  fits <- reconcilethief(frc$fitted, comb, frc$mse, frc$residuals,
    returnall=FALSE, aggregatelist=aggregatelist)

  # Truncate to h forecasts
  tspb <- tsp(bts)
  bts <- ts(bts[1:h], start=tspb[1], frequency=tspb[3])

  # Construct forecast object to return
  out <- structure(list(x=y, mean=bts, fitted=fits, residuals=y-fits),
    class='forecast')
  out$method <- paste("THieF-",toupper(usemodel),sep="")
  return(out)
}

#-------------------------------------------------
th.forecast <- function(aggy, h=NULL, usemodel, forecastfunction, ...)
{
# Produce forecasts for multiple temporal aggregation levels with predefined models
#
# Inputs:
#   aggy        List of time series to be forecast
#   h           Forecast horizon. Default is m*2
#   usemodel    Model used for forecasting each aggregation level
#
# Outputs:
#   frc         Uncombined temporal hierarchy forecasts
#   fitted      Uncombined temporal hierarchy fitted values
#   residual    Uncombined temporal hierarchy residuals
#   mse         Mean squared errors for each temporal aggregation level
#   mseh        Mean squared errors per horizon for each temporal aggregation level
#
# Example:
#   aggy <- tsaggregates(USAccDeaths)
#   th.forecast(aggy,h=20)
#
# Nikolaos Kourentzes and Rob J Hyndman

  # Calculate forecast horizons
  n.k <- length(aggy)
  freq <- unlist(lapply(aggy, frequency))
  m <- max(freq)
  AL <- m/freq
  H <- m/AL*ceiling(h/m)

  # Initialise
  frc <- fitted <- resid <- mseh <- vector("list",n.k)
  mse <- vector("numeric",n.k)

  # Model estimation and forecasts
  for (k in 1:n.k)
  {
    temp <- th.forecast.loop(aggy[[k]], H[[k]], usemodel, forecastfunction, ...)
    frc[[k]] <- temp$frc
    fitted[[k]] <- temp$fitted
    resid[[k]] <- temp$resid
    mse[k] <- temp$mse
    mseh[[k]] <- temp$mseh
  }

  # Return forecasts
  #names(frc) <- names(fitted) <- names(resid) <- names(mseh) <- names(mse) <- names(aggy)

  return(list(forecast=frc,fitted=fitted,residuals=resid,mse=mse,mseh=mseh))
}

#-------------------------------------------------
th.forecast.loop <- function(y, h, usemodel, forecastfunction, ...){
# Loop for model forecasts
# Produce forecasts and other information for univariate series y
# Forecast horizon h
# Return list of forecasts, fitted values, residuals, mse, etc.

  m <- frequency(y)

  if(!is.null(forecastfunction))
  {
    fc <- forecastfunction(y, h, ...)
    # Check if PI returned.
    if(!is.null(fc$lower))
    {
      # Check only 80% PI is saved
      if(is.element(80,fc$level))
        fc$lower <- fc$lower[,fc$level==80]
      else
        fc$lower <- NULL
    }
  }
  else if (usemodel=="ets")
  {
    fit <- try(forecast::ets(y, ...), silent=TRUE)
    # Do something simpler if ets model doesn't work
    # This is only necessary for versions of the package before 22 Aug 2016 (7.2beta)
    if(is.element("try-error",class(fit)))
    {
      # Use HW if there is enough data and the data is seasonal
      if(frequency(y) > 1L & length(y) >= 2*frequency(y))
        fit <- fc <- forecast::hw(y, h=h, level=80,
                        initial='simple', alpha=0.2, beta=0.1, gamma=0.01)
      else # Otherwise just use Holt's
        fit <- fc <- forecast::holt(y, h=h, level=80,
                        initial='simple', alpha=0.2, beta=0.1)
    }
    else
      fc <- forecast::forecast(fit, h=h, level=80)
  }
  else if(usemodel=="arima")
  {
    fit <- forecast::auto.arima(y, ...)
    fc <- forecast::forecast(fit, h=h, level=80)
  }
  else if (usemodel == "theta")
  {
    fc <- forecast::thetaf(y, h=h, level=80)
  }
  else
  {
    # Seasonal naive forecast
    if(usemodel=='snaive')
      fc <- forecast::snaive(y, h=h, level=80)
    # Naive forecast
    else
      fc <- forecast::naive(y, h=h, level=80)
  }

  # Calculate fitted residuals and MSE
  fitted <- fitted(fc)
  if(is.null(fitted) & !is.null(fc$residuals))
    fitted <- y - fc$residuals
  if(!is.null(fitted))
    resid <- y - fitted
  else
    resid <- NULL
  mse <- mean(resid^2, na.rm=TRUE)

  # Only keep full years of fitted values and residuals
  # Remove partial year from start
  tspy <- tsp(y)
  fullyears <- trunc(length(y) / m)
  fitted <- ts(utils::tail(fitted, fullyears*m), frequency=m, end=tspy[2])
  resid <- ts(utils::tail(resid, fullyears*m), frequency=m, end=tspy[2])

  # Compute MSE and MSEH
  if(!is.null(fc$lower))
    mseh <- ((fc$mean - fc$lower)/stats::qnorm(0.9))^2
  else
    mseh <- NULL

  return(list(frc=fc$mean,mse=mse,mseh=mseh,fitted=fitted,resid=resid))
}
